wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/jpa/JpaUserDevfileDao.java

```diff
@@ -0,0 +1,420 @@
+
+  @Transactional
+  protected Optional<UserDevfileImpl> doUpdate(UserDevfileImpl update) {
+    final EntityManager manager = managerProvider.get();
+    if (manager.find(UserDevfileImpl.class, update.getId()) == null) {
+      return Optional.empty();
+    }
+    UserDevfileImpl merged = manager.merge(update);
+    manager.flush();
+    return Optional.of(merged);
+  }
+
+  @Transactional(rollbackOn = {RuntimeException.class, ServerException.class})
+  protected void doRemove(String id) throws ServerException {
+    final EntityManager manager = managerProvider.get();
+    final UserDevfileImpl devfile = manager.find(UserDevfileImpl.class, id);
+    if (devfile != null) {
+      eventService
+          .publish(new BeforeDevfileRemovedEvent(new UserDevfileImpl(devfile)))
+          .propagateException();
+      manager.remove(devfile);
+      manager.flush();
+    }
+  }
+
+  public static class UserDevfileSearchQueryBuilder {
+    protected EntityManager entityManager;
+    protected int maxItems;
+    protected int skipCount;
+    protected String filter;
+    protected Map<String, String> params;
+    protected String order;
+
+    public UserDevfileSearchQueryBuilder(EntityManager entityManager) {
+      this.entityManager = entityManager;
+      this.params = new HashMap<>();
+    }
+
+    public static UserDevfileSearchQueryBuilder newBuilder(EntityManager entityManager) {
+      return new JpaUserDevfileDao.UserDevfileSearchQueryBuilder(entityManager);
+    }
+
+    public UserDevfileSearchQueryBuilder withMaxItems(int maxItems) {
+      this.maxItems = maxItems;
+      return this;
+    }
+
+    public UserDevfileSearchQueryBuilder withSkipCount(int skipCount) {
+      this.skipCount = skipCount;
+      return this;
+    }
+
+    public UserDevfileSearchQueryBuilder withFilter(List<Pair<String, String>> filter) {
+      if (filter == null || filter.isEmpty()) {
+        this.filter = "";
+        return this;
+      }
+      final StringJoiner matcher = new StringJoiner(" AND ", " WHERE ", " ");
+      int i = 0;
+      for (Pair<String, String> attribute : filter) {
+
+        final String parameterName = "parameterName" + i++;
+        if (attribute.second.startsWith("like:")) {
+          params.put(parameterName, attribute.second.substring(5));
+          matcher.add("userdevfile." + attribute.first + " LIKE :" + parameterName);
+        } else {
+          params.put(parameterName, attribute.second);
+          matcher.add("userdevfile." + attribute.first + " = :" + parameterName);
+        }
+      }
+      this.filter = matcher.toString();
+      return this;
+    }
+
+    public UserDevfileSearchQueryBuilder withOrder(List<Pair<String, String>> order) {
+      if (order == null || order.isEmpty()) {
+        this.order = "";
+        return this;
+      }
+      final StringJoiner matcher = new StringJoiner(", ", " ORDER BY ", " ");
+      order.forEach(pair -> matcher.add("userdevfile." + pair.first + " " + pair.second));
```

Comment:
Same thing as with `attribute.first` in the `withFilter` method. We should make sure (as much as comments allow) to not shoot ourselves in the foot in the future by allowing unsanitized string in the query.